drivers/SmartThings/zigbee-thermostat/src/popp/common.lua

```diff
@@ -0,0 +1,222 @@
+}
+common.EXTERNAL_OPEN_WINDOW_DETECTION_ID = 0x4003
+
+local SUPPORTED_MODES = {
+  ThermostatMode.thermostatMode.off.NAME,
+  ThermostatMode.thermostatMode.heat.NAME,
+  ThermostatMode.thermostatMode.eco.NAME
+}
+
+-- Helpers
+
+-- has member check function
+local has_member = function(haystack, needle)
+  for _, value in ipairs(haystack) do
+    if (value == needle) then
+      return true
+    end
+  end
+  return false
+end
+
+-- Turn switch 'on' if its state is 'off'
+local turn_switch_on = function(driver, device)
+  -- turn thermostat ventile on
+  if device:get_latest_state("main", Switch.ID, Switch.switch.NAME) == "off" then
+    common.switch_handler_factory('on')(driver, device)
+  end
+end
+
+-- Handlers
+
+-- Internal window open detection handler
+common.window_open_detection_handler = function(driver, device, value, zb_rx)
+  device:emit_event(TemperatureAlarm.temperatureAlarm(common.WINDOW_OPEN_DETECTION_MAP[value.value]))
+end
+
+-- Switch handler
+common.switch_handler_factory = function(switch_state)
+  return function(driver, device, cmd)
+    local get_cmd = switch_state or cmd.command
+    local external_window_open
+
+    if get_cmd == 'on' then
+      external_window_open = false
+    elseif get_cmd == 'off' then
+      external_window_open = true
+    end
+
+    device:send(cluster_base.write_manufacturer_specific_attribute(device, common.THERMOSTAT_CLUSTER_ID,
+      common.EXTERNAL_OPEN_WINDOW_DETECTION_ID,
+      common.MFG_CODE, data_types.Boolean, external_window_open))
+  end
+end
+
+-- Custom setpoint command handler
+common.setpoint_cmd_handler = function(driver, device, cmd)
+  local payload
+  local mode = cmd.args.mode
+
+  if has_member(SUPPORTED_MODES, mode) then
+
+    -- fetch last_setpointTemp
+    last_setpointTemp = device:get_field("last_setpointTemp")
+
+    if last_setpointTemp == nil then
+      last_setpointTemp = device:get_latest_state("main", ThermostatHeatingSetpoint.ID, ThermostatHeatingSetpoint.heatingSetpoint.NAME) or 21 -- thermostat default temperature
+    end
+
+    -- prepare setpoint for correct 4 char dec format
+    last_setpointTemp = math.floor(last_setpointTemp * 100)
+
+    -- convert setpoint value into bytes e.g. 25.5 -> 2550 -> \x09\xF6 -> \xF6\x09
+    local s = string.format("%04X", tostring(last_setpointTemp))
+    local p2 = tonumber(string.sub(s, 3, 4), 16)
+    local p3 = tonumber(string.sub(s, 1, 2), 16)
+    local type = nil
+
+    if mode == ThermostatMode.thermostatMode.heat.NAME then
+      -- Setpoint type "1": the actuator will make a large movement to minimize reaction time to UI
+      type = 0x01
+    elseif mode == ThermostatMode.thermostatMode.eco.NAME then
+      -- Setpoint type "0": the behavior will be the same as setting the attribute "Occupied Heating Setpoint" to the same value
+      type = 0x00
+    end
+
+    -- send thermostat setpoint command
+    payload = string.char(type, p2, p3)
```

Comment:
This will throw an exception if `type` is `nil`, which would be the case if the `mode` is `off`.